types/prefs/prefs_test.go

```diff
@@ -0,0 +1,670 @@
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/netip"
+	"reflect"
+	"testing"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"github.com/google/go-cmp/cmp"
+	"tailscale.com/types/views"
+)
+
+//go:generate go run tailscale.com/cmd/viewer --tags=test --type=TestPrefs,TestBundle,TestValueStruct,TestGenericStruct,TestPrefsGroup
+
+type TestPrefs struct {
+	Int32Item   Item[int32]  `json:",omitzero"`
+	UInt64Item  Item[uint64] `json:",omitzero"`
+	StringItem1 Item[string] `json:",omitzero"`
+	StringItem2 Item[string] `json:",omitzero"`
+	BoolItem1   Item[bool]   `json:",omitzero"`
+	BoolItem2   Item[bool]   `json:",omitzero"`
+	StringSlice List[string] `json:",omitzero"`
+	IntSlice    List[int]    `json:",omitzero"`
+
+	AddrItem Item[netip.Addr] `json:",omitzero"`
+
+	StringStringMap Map[string, string]  `json:",omitzero"`
+	IntStringMap    Map[int, string]     `json:",omitzero"`
+	AddrIntMap      Map[netip.Addr, int] `json:",omitzero"`
+
+	// Bundles are complex preferences that usually consist of
+	// multiple parameters that must be configured atomically.
+	Bundle1 Item[*TestBundle]             `json:",omitzero"`
+	Bundle2 Item[*TestBundle]             `json:",omitzero"`
+	Generic Item[*TestGenericStruct[int]] `json:",omitzero"`
+
+	BundleList StructList[*TestBundle] `json:",omitzero"`
+
+	StringBundleMap StructMap[string, *TestBundle]     `json:",omitzero"`
+	IntBundleMap    StructMap[int, *TestBundle]        `json:",omitzero"`
+	AddrBundleMap   StructMap[netip.Addr, *TestBundle] `json:",omitzero"`
+
+	// Group is a nested struct that contains one or more preferences.
+	// Each preference in a group can be configured individually.
+	// Preference groups should be included directly rather than by pointers.
+	Group TestPrefsGroup `json:",omitzero"`
+}
+
+// MarshalJSONV2 implements [jsonv2.MarshalerV2].
+func (p TestPrefs) MarshalJSONV2(out *jsontext.Encoder, opts jsonv2.Options) error {
+	// The testPrefs type shadows the TestPrefs's method set,
+	// causing jsonv2 to use the default marshaler and avoiding
+	// infinite recursion.
+	type testPrefs TestPrefs
+	return jsonv2.MarshalEncode(out, (*testPrefs)(&p), opts)
+}
+
+// UnmarshalJSONV2 implements [jsonv2.UnmarshalerV2].
+func (p *TestPrefs) UnmarshalJSONV2(in *jsontext.Decoder, opts jsonv2.Options) error {
+	// The testPrefs type shadows the TestPrefs's method set,
+	// causing jsonv2 to use the default unmarshaler and avoiding
+	// infinite recursion.
+	type testPrefs TestPrefs
+	return jsonv2.UnmarshalDecode(in, (*testPrefs)(p), opts)
+}
+
+// MarshalJSON implements [json.Marshaler].
+func (p TestPrefs) MarshalJSON() ([]byte, error) {
+	return jsonv2.Marshal(p) // uses MarshalJSONV2
+}
+
+// UnmarshalJSON implements [json.Unmarshaler].
+func (p *TestPrefs) UnmarshalJSON(b []byte) error {
+	return jsonv2.Unmarshal(b, p) // uses UnmarshalJSONV2
+}
```

Comment:
How would `TestPrefs` marshal if you didn't have these methods?

Reply:
[Slower](https://www.youtube.com/watch?feature=shared&t=688&v=avilmOcHKHE) and without omitting unconfigured preferences (`encoding/json` does not support the `omitzero` tag option yet).